src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -263,7 +263,30 @@ private <T extends SyntaxElement> T parse(Iterator<? extends SyntaxElementInfo<?
 								x = x2;
 							}
 							T t = info.c.newInstance();
-							if (t.init(res.exprs, i, getParser().getHasDelayBefore(), res)) {
+							ParserInstance parser = getParser();
+							List<Class<? extends Structure>> usableStructures = t.getUsableStructures();
+							if (!usableStructures.isEmpty()) {
+								if (parser.getCurrentStructure() == null) {
+									Skript.error("'" + res.expr + "' cannot be used outside of the allowed structures.");
+									continue;
+								} else if (usableStructures.stream().noneMatch(parser::isCurrentStructure)) {
+									Skript.error("'" + res.expr + "' cannot be used in " + Utils.a(parser.getCurrentStructure().toString(null, false)) + " structure.");
+									continue;
+								}
+							}
+							List<Class<? extends Section>> usableSections = t.getUsableSections();
+							if (!usableSections.isEmpty()) {
+								List<TriggerSection> sections = parser.getCurrentSections();
+								TriggerSection current = sections.isEmpty() ? null : sections.get(sections.size() - 1);
+								if (sections.isEmpty()) {
+									Skript.error("'" + res.expr + "' cannot be used outside of the allowed sections.");
+									continue;
+								} else if (usableSections.stream().noneMatch(parser::isCurrentSection)) {
+									Skript.error("'" + res.expr + "' cannot be used in '" + Utils.a(current.toString(null, false)) + "' section.");
+									continue;
+								}
+							}
+							if (t.init(res.exprs, i, parser.getHasDelayBefore(), res)) {
```

Comment:
It would be appreciate to use proper variable names (i, t, res)

Reply:
I could change this however this was just the original file. I wouldnt want to go about changing all the names which i feel is unrelated to this PR, i feel a variable formatting change PR for the whole file would be more viable for this, unless more disagree and would like me to do this ! LMK tho c:

Reply:
It's indeed out of scope of this PR. Maybe one day later c: